### ISSUE TITLE
Support const parameters in type dependent paths

### ISSUE THREAD

**Issue by lcnr** [rust-lang/compiler-team#304]
# What is this issue?

This is a **major change proposal**, which means a proposal to make a notable change to the compiler -- one that either alters the architecture of some component, affects a lot of people, or makes a small but noticeable public change (e.g., adding a compiler flag). You can read more about the MCP process on https://forge.rust-lang.org/.

**This issue is not meant to be used for technical discussion. There is a Zulip stream for that. Use this issue to leave procedural comments, such as volunteering to review, indicating that you second the proposal (or third, etc), or raising a concern that you would like to be addressed.**

# MCP Checklist

* [x] MCP **filed**. Automatically, as a result of filing this issue:
  * The @rust-lang/wg-prioritization group will add this to the triage meeting agenda so folks see it.
  * A Zulip topic in the stream `#t-compiler/major changes` will be created for this issue.
* [ ] MCP **seconded**. The MCP is "seconded" when a compiler team member or contributor issues the `@rustbot second` command. This should only be done by someone knowledgable with the area -- before seconding, it may be a good idea to cc other stakeholders as well and get their opinion.
* [ ] **Final comment period** (FCP). Once the MCP is approved, the FCP begins and lasts for 10 days. This is a time for other members to review and raise concerns -- **concerns that should block acceptance should be noted as comments on the thread**, ideally with a link to Zulip for further discussion.
* [ ] MCP **Accepted**. At the end of the FCP, a compiler team lead will review the comments and discussion and decide whether to accept the MCP.
  * At this point, the `major-change-accepted` label is added and the issue is closed. You can link to it for future reference.

# TL;DR

wip PR implementing this change: https://github.com/rust-lang/rust/pull/71154

Support explicit generic const parameters in type dependend paths. Prevent cycle errors by bundling the `DefId` of a generic argument with the `DefId` of its generic parameter in relevant queries.

# Links and Details

We want to be able to annotate method calls with explicit generic const arguments, e.g. `x.array_chunks::<{1 << 4}>()`.

When type-checking a function containing the above expression, we have to evaluate `{1 << 4}`. This requires us to generate mir for this `AnonConst` which requires it be to first be type-checked. As the type of a const argument depends on its parameter, this requires us to know the type of `N` in `<_>::array_chunks::<N>`. To know the type of `N` we once again have to type-check the function containing this expression, resulting in a cycle error. 

As const arguments only need to be evaluated once the corresponding generic parameter is known, we can prevent this cycle by passing the corresponding parameter into all queries which potentially cause cycles during type-checking.

This is done using a new struct `WithOptParam`, which stores the `DefId` of the relevant parameter next to the `(Local)DefId` of the actual argument.

`WithOptParam` can be created using `tcx.with_opt_param(def_id)`, which is now fairly frequently used throughout the compiler.

Queries now taking `WithOptParam` as an argument:
`mir_const_qualif`, `mir_built`, `mir_const`, `mir_validated`, `promoted_mir`, `unsafety_check_result`, `optimized_mir`,  `typeck_tables_of`, `mir_borrowck` and `resolve_instance`.

As `WithOptParam` should always contain the correct corresponding parameter by construction, any missed queries may only lead to cycle errors and should not be able to cause miscompilations.

# Mentors or Reviewers

probably mentored by @varkor and @eddyb
